src/rebar_compiler_dag.erl

```diff
@@ -237,41 +237,58 @@ propagate_stamps(G) ->
 
 
 %% @doc Return the reverse sorting order to get dep-free apps first.
-%% -- we would usually not need to consider the non-source files for the order to
-%% be complete, but using them doesn't hurt.
+
 compile_order(_, AppDefs) when length(AppDefs) =< 1 ->
     [Name || {Name, _Path} <- AppDefs];
 compile_order(G, AppDefs) ->
     %% Build a digraph for following topo-sort, and populate
     %%  FileToApp map as a side effect for caching
     AppDAG = digraph:new([acyclic]), % ignore cycles and hope it works
-    lists:foldl(
-        fun(E, FileToApp) ->
-            case digraph:edge(G, E) of
-                {_, _, _, artifact} ->
-                    %% skip artifacts, they don't affect compile order
-                    FileToApp;
-                {_, V1, V2, _} ->
-                    %% V1 depends on V2, and we know V1 is in our repo,
-                    %%  otherwise this edge would've never appeared
-                    {App, FTA1} = find_and_cache(V1, AppDefs, FileToApp),
-                    case find_and_cache(V2, AppDefs, FTA1) of
-                        {[], FTA2} ->
-                            %% dependency on a file outside of the repo
-                            FTA2;
-                        {App, FTA2} ->
+    OneDep =
+        fun(V1, V2, Cache) ->
+            %% First resolve the file we depend on so that we can shortcut resolution
+            %% If it is a file outside the repo.
+            case resolve_file_to_app(V2, AppDefs, Cache) of
+                {[], Cache1} ->
+                    %% dependency on a file outside of the repo
+                    Cache1;
+                {App, Cache1} ->
+                    case resolve_file_to_app(V1, AppDefs, Cache1) of
+                        {App, Cache2} ->
                             %% dependency within the same app
-                            FTA2;
-                        {AnotherApp, FTA2} ->
+                            Cache2;
+                        {AnotherApp, Cache2} when AnotherApp =/= [] ->
                             %% actual dependency
                             %% unfortunately digraph has non-functional API depending on side effects
                             digraph:add_vertex(AppDAG, App), %% ignore errors for duplicate inserts
                             digraph:add_vertex(AppDAG, AnotherApp),
-                            digraph:add_edge(AppDAG, App, AnotherApp),
-                            FTA2
+                            digraph:add_edge(AppDAG, AnotherApp, App),
+                            Cache2;
+                        {[], Cache2} ->
```

Comment:
move this clause up so all short-circuits are in a row, and we can avoid a guard check as well.

Reply:
Done